File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-4.cpp
class Solution {
private:
    vector<vector<int>> ret; 
    void dfs(vector<vector<int>>& heights, int r, int c, int& sum){
        if (sum == 3)   return;
        if ( r - 1 < 0 || c - 1 < 0){ //pacific
            if (sum == 2 || sum == 0){  //either atlantic visited or not visited it
                sum++;
            }
        }if (r + 1 >= heights.size() || c + 1 >= heights[0].size()){
            if (sum == 0 || sum == 1){  //either visited pacific or not visited it
                sum++;sum++;
            }
        }if (sum == 3) return;

        if (r + 1 < heights.size())
            if(heights[r+1][c] != -1 && heights[r+1][c] <= heights[r][c]){
                int temp = heights[r][c];
                heights[r][c] = -1;
                dfs(heights, r + 1, c, sum);
                heights[r][c] = temp;
            }
        if (r - 1 >= 0) 
            if(heights[r-1][c] != -1 && heights[r-1][c] <= heights[r][c]){
                int temp = heights[r][c];
                heights[r][c] = -1;
                dfs(heights, r - 1, c, sum);
                heights[r][c] = temp;
            }
        if (c + 1 < heights[0].size())
            if(heights[r][c+1] != -1 && heights[r][c+1] <= heights[r][c]){
                int temp = heights[r][c];
                heights[r][c] = -1;
                dfs(heights, r, c + 1, sum);
                heights[r][c] = temp;
            }
        if (c - 1 >= 0)
            if(heights[r][c-1] != -1 && heights[r][c-1] <= heights[r][c]){
                int temp = heights[r][c];
                heights[r][c] = -1;
                dfs(heights, r, c - 1, sum);
                heights[r][c] = temp;
            }
    }

public:
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        for (int r = 0 ; r < heights.size() ; r++){
            for (int c = 0 ; c < heights[0].size() ; c++){
                int sum = 0;
                dfs(heights, r, c, sum);
                if (sum == 3){
                    ret.push_back({r,c});
                }
            }
        }return ret;
    }
};

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-4.cs
public class Solution {
    public bool CanPartition(int[] nums) {
        if (nums.Sum() % 2 != 0) return false;
        // bool[] visited = new bool[nums.Length];
        // Array.Fill(visited, false);
        Dictionary<(int, int), bool> dp = new Dictionary<(int, int), bool>();
        
        return dfs(nums, dp, 0, nums.Sum() / 2);
    }

    bool dfs(int[] nums, Dictionary<(int, int), bool> dp, int idx, int target){
        if (target == 0) return true;
        if (idx == nums.Length || target < 0) return false;
        if (dp.ContainsKey((idx, target))) return dp[(idx, target)];

        //skip current number
        bool ret = dfs(nums, dp, idx + 1, target);

        //take current number
        ret = ret || dfs(nums, dp, idx + 1, target - nums[idx]);

        dp[(idx, target)] = ret;
        return ret;
    }
}
